main.js

```diff
@@ -0,0 +1,313 @@
+    // pad start
+    const data = new Uint8Array(bitLength / 8);
+    data.set(bytes, data.length - bytes.length);
+    return data;
+  }
+  // trim start, ensure trimmed data is zero
+  const start = (length - bitLength) / 8;
+  if(bytes.subarray(0, start).some(d => d !== 0)) {
+    throw new Error(
+      `Data length greater than ${bitLength} bits.`);
+  }
+  return bytes.subarray(start);
+}
+
+const _log2_16 = 4;
+function _base16Encoder({bytes, idEncoder}) {
+  let encoded = bytesToHex(bytes);
+  if(idEncoder.encoding === 'base16upper') {
+    encoded = encoded.toUpperCase();
+  }
+  if(idEncoder.fixedLength) {
+    const fixedBitLength = _calcDataBitLength({
+      bitLength: bytes.length * 8,
+      maxLength: idEncoder.fixedBitLength
+    });
+    const wantLength = Math.ceil(fixedBitLength / _log2_16);
+    // pad start with 0s
+    return encoded.padStart(wantLength, '0');
+  } else {
+    return encoded;
+  }
+}
+
+const _log2_58 = Math.log2(58);
+function _base58Encoder({bytes, idEncoder}) {
+  const encoded = base58encoder(bytes);
+  if(idEncoder.fixedLength) {
+    const fixedBitLength = _calcDataBitLength({
+      bitLength: bytes.length * 8,
+      maxLength: idEncoder.fixedBitLength
+    });
+    const wantLength = Math.ceil(fixedBitLength / _log2_58);
+    // pad start with 0s (encoded as '1's)
+    return encoded.padStart(wantLength, '1');
+  } else {
+    return encoded;
+  }
+}
+
+export class IdGenerator {
+  /**
+   * Creates a new IdGenerator instance.
+   *
+   * An IdGenerator generates an array of id bytes.
+   *
+   * @param {Number} [bitLength] - number of bits to generate. May also be an
+   *   object with a bitLength property.
+   *
+   * @return {IdGenerator}
+   */
+  constructor(bitLength) {
+    // handle as object property
+    if(typeof bitLength === 'object') {
+      bitLength = bitLength.bitLength;
+    }
+    this.bitLength = _calcOptionsBitLength({
+      // default to 128 bits / 16 bytes
+      defaultLength: 128,
+      // TODO: allow any bit length
+      minLength: 8,
+      bitLength,
+    });
+  }
+
+  /**
+   * Generate random id bytes.
+   *
+   * @return {Uint8Array}
+   */
+  async generate() {
+    const buf = new Uint8Array(this.bitLength / 8);
+    await getRandomBytes(buf);
+    return buf;
+  }
+}
+
+export class IdEncoder {
+  /**
+   * Creates a new IdEncoder instance.
+   *
+   * An IdEncoder encodes an array of id bytes into a specific encoding.
+   *
+   * @param {string} [encoding='base58'] - encoding format.
+   * @param {boolean} [fixedLength=false] - true to ensure fixed output length.
+   * @param {Number} [fixedBitLength] - fixed output bit length or 0 to base on
+   *   input byte size.
+   * @param {boolean} [multibase=false] - use multibase encoding.
+   *
+   * @return {IdEncoder}
+   */
+  constructor({
+    encoding = 'base58',
+    fixedLength = false,
+    fixedBitLength,
+    multibase = false
+  } = {}) {
+    switch(encoding) {
+      case 'hex':
```

Comment:
Douglas Crockford hates switch statements and this fallthrough pattern in particular.  Do we need these aliases in the first place?

Reply:
Switch statements are fine.  Who cares what he hates?  I had started with an encoding map of name to handler, but ran out of time and used this simpler thing for now.  We could stick with just multibase names.  But I figured 'hex' and 'base58' would be more recognizable.  Should I remove them?

Reply:
Just keep them for now.